<a href="https://colab.research.google.com/github/johanhoffman/DD2363_VT23/blob/main/template-report-lab-X.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Lab 2: Iterative methods**
**Matteus Berg**

# **Abstract**




This lab reports implements the newton iteration for nonlinear scalar equations, the jacobi iteration and the gauss-seidel iteration. The results show that the algorithms that were implemented work as expected.

#**About the code**



```
# Detta formeras som kod
```

The lab report is based on a template given by Johan Hoffman.

'KTH Royal Institute of Technology, Stockholm, Sweden.'

# **Set up environment**

To have access to the neccessary modules you have to run this cell. If you need additional modules, this is where you add them.

In [ ]:
# Load neccessary modules.
from google.colab import files

import time
import numpy as np

#try:
#    from dolfin import *; from mshr import *
#except ImportError as e:
#    !apt-get install -y -qq software-properties-common
#    !add-apt-repository -y ppa:fenics-packages/fenics
#    !apt-get update -qq
#    !apt install -y --no-install-recommends fenics
#    from dolfin import *; from mshr import *

#import dolfin.common.plotting as fenicsplot

from matplotlib import pyplot as plt
from matplotlib import tri
from matplotlib import axes
from mpl_toolkits.mplot3d import Axes3D

# **Introduction**

For the Jacobi and Gauss-Siedel iterations, we want to solve the following linear equation $Ax = b$. Where $A{\in}C^{nxn}$ and $b,x{\in}C^n$. We want to find x. This could be done by just calculating the inverse of $A$, but if this for some reason is not feasible, x can be calculated by an iterative method. Gauss-Seidel and Jacobi are two different such iterative methods.

Newton's method for nonlinear scalar equations solves the equation $f(x) = 0$. Where f(x) is a nonlinear scalar function. It does this by iteratively converging on the correct solution for x.

The problem investigated in this report is implementing these three methods.

For all methods, I decided to have the tolerance be 0.05.



# **Method**

The three algorithms are written below in code. Tests are formulated for each method which can be run under the results.

The implementation for the Jacobi and Gauss-Seidel iterations are implemented in accordance with equation 7.12 in the book.
$x^{k+1} = x^{(k)} + {\alpha}r^{(k)}$

Where ${\alpha} = 1$ and $r^{(k)}$ is equal to
* $L^{-1}(b - Ax^{(k)})$ for Gauss-Seidel
* $D^{-1}(b - Ax^{(k)})$ for Jacobi

Where $D = diag(A)$ and $L$ is the lower triangular matrix of $A$.

In [ ]:

# Newton's method for scalar nonlinear equation f(x)=0
def newtons_method(f, fprim, x0):
    x = x0
    while(abs(f(x)) > 0.05):
        x = x - f(x)/fprim(x)
    return x

# Gauss seidel iteration
def gauss_seidel(A: np.array, b: np.array):
    x = np.zeros((A.shape[0], 1))
    r = np.zeros((A.shape[0], 1))
    L = np.zeros(A.shape)
    L_inv = np.zeros(A.shape)
    I = np.zeros(A.shape)
    # generate L, L inverse and identity matrix
    for i in range(A.shape[0]):
        I[i,i] = 1
        for j in range(A.shape[1] - (A.shape[1]-i-1)):
            L[i,j] = A[i,j]
    L_inv = np.linalg.inv(L)

    while True:
        r = np.matmul(A, x)
        r = np.subtract(b, r)
        r = np.matmul(L_inv, r)
        x = np.add(x, r)
        if (not (np.linalg.norm(r)/np.linalg.norm(b)) > 0.05):
            break
    return x

# Jacobi iteration
# We use the following formula to calculate the iteration matrix and vector
# M_j = (I - D_inv * A)
# c = D_inv * b
def jacobi(A: np.array, b: np.array):
    x = np.zeros((A.shape[0], 1))
    r = np.zeros((A.shape[0], 1))
    D = np.zeros(A.shape)
    D_inv = np.zeros(A.shape)
    I = np.zeros(A.shape)
    # generate D, D inverse and identity matrix
    for i in range(A.shape[0]):
        D[i,i] = A[i,i]
        D_inv[i,i] = 1/A[i,i]
        I[i,i] = 1

    while True:
        r = np.matmul(A, x)
        r = np.subtract(b, r)
        r = np.matmul(D_inv, r)
        x = np.add(x, r)
        if (not (np.linalg.norm(r)/np.linalg.norm(b)) > 0.05):
            break
    return x


# **Results**

In [ ]:
# Newton's method scalar nonlinear equation
print("Newton: ", newtons_method(np.sin, np.cos, 1), "\n")


A = np.array([[4,2],[0,8]])
b = np.array([[10],[5]])
# Gauss Seidel
print("Gauss:\n", gauss_seidel(A,b), "\n")

# Jacobi
print("Jacobi:\n", jacobi(A,b), "\n")


Newton:  -9.572191932508134e-05 

Gauss:
 [[2.1875]
 [0.625 ]] 

Jacobi:
 [[2.1875]
 [0.625 ]] 



Here above we see test input to the three functions.

For solving the equation sin(x) = 0, newton's method generates a value really close to 0. This means that the iteration converges to the correct solution.

For solving the following linear equation
\begin{gather}
  \begin{bmatrix}
    4 & 2\\
    0 & 8
  \end{bmatrix}
  \begin{bmatrix}
    x_1\\
    x_2
  \end{bmatrix}
  =
  \begin{bmatrix}
    10\\
    5
  \end{bmatrix}
\end{gather}

The Jacobi iteration and Gauss-Seidel iteration generates the solution
$x = [2.1875, 0.625]^T$, which is correct.

# **Discussion**


In conclusion, three algorithms have been implmented. These where
* Newton's iteration for nonlinear scalar equations
* Jacobi iteration
* Gauss Seidel iteration

All three algorithms worked as expected.